### DISCUSSION TITLE
what's the recommended tps for SBOM upload

### DISCUSSION THREAD

**Discussion by andreeaButerchi** [DependencyTrack/dependency-track#3818] [Q&A] [answered]
Hello,

We've trying to use DT in a batch mode: meaning we upload a lot of SBOM files and then we wait for the processing to be done... but I think we're causing a bottleneck :( 
I did not manage to find any recommendation regarding the SBOM ingestion ratio...
are there any recommendations/thresholds/benchmarks? Or the container resources is the limit?
knowing that an average SBOM is 6-700Kb and we're at a 5 TPS (sbom uploads)... but we can see the processing getting longer & longer & longer :( 
we're considering uploading something like 20 and send another 20 only when the 1st ones are done... Knowing that 20 is just an example... we'll try to figure out the magic number later. Just wanted to know if it might be a good idea...
or maybe we're doing something wrong 

Thank you very much for your help!

**Comment by nscuro** [comment, marked as answer]
It's very hard if not impossible to provide one-size-fits-all recommendations.

The symptom you're experiencing is one of the reasons we started working on Hyades, which eventually will become Dependency-Track v5. We wrote about this particular issue here: https://github.com/DependencyTrack/hyades/blob/main/WTF.md#why

Still, you *can* improve the current situation, but you will need to start monitoring application metrics, otherwise you'll be blind to where the actual limitation lies. We support exposition of Prometheus metrics, and provide an example Grafana dashboard that you can use: https://docs.dependencytrack.org/getting-started/monitoring/#grafana-dashboard

For a start, have a look at the eventing metrics:

![image](https://github.com/DependencyTrack/dependency-track/assets/5693141/cfeb8308-3e1f-4d96-84fd-e9135c7e8044)

You'll likely see high numbers of *Events Queued* and a constantly saturated *Active Worker Threads*. Depending on how large you worker pool is at the moment, you can try increasing it and see if it helps:

https://github.com/DependencyTrack/dependency-track/blob/b3f96ea25f59c0257e9d7d2bc849d31120a845af/src/main/resources/application.properties#L3-L19

For example, you can pass the following environment variable to increase the worker threads to 32:

```
ALPINE_WORKER_THREADS=32
```

You should be catious with this though, as more concurrency can cause more load / contention on the database as well. Have an eye on the CPU / memory and database metrics on the dashboard as well.

Separately, in DT v4.11 we introduced a new way of processing uploaded BOMs, which is significantly faster / more efficient. You can enable it under *Configuration* -> *Experimental* -> *BOM Processing V2* in the settings. It will become the default in DT v4.12.

**Reply by andreeaButerchi** [reply]
Hello,

Thank you very very very much for your quick reply and complete details :) 
We're working on doing the improvements... well to be honest we've done some just after your answer :D The results are quite promissing, but we want to double check!!
Anyhow thanks a lot and we'll keep you posted on the results :)

**Reply by nekhtan** [reply]
Hello @nscuro ,

I'm working with Andreea on the integration of ODT. We tried quite a few things and so far nothing seems to do the trick.

We are limited by the kube cluster node's memory: it has a total of 32GB, but roughly 16GB are available (depending on the time of day due to other batches).
So basically when we increase the worker threads count and the memory, the server eventually crashes due to the lack of memory on the node (pure hypothesis since we can't find any log whatsoever regarding why it crashed - the pod is dead so Kube logs are gone, and nothing unusual on ODT side).
This leads me to a first question (I think I know the answer): is the queue lost or do you maintain state in database ?

Increasing the memory to more than 16GB will be tricky due to our infra limitations. So we decreased it to 12GB and reduced the threads count accordingly (tried 20, or playing with the multiplier -> turns out it's based on the cpu limit, right ?). 

Nothing worked, we disabled the BOM validation, increased the non-transactional pool size (hitting the threshold), increased the portfolio/vuln metrics updates to 6h, and lately we were considering disabling the Maven repositories (metadata analysis - looks like it's only fetching the latest version). Basically we're trying to have a processing as light as possible.

But looking at the graphs, we wonder: can we configure the datanucleus max size ? The docs mention that we can update its type but nothing deals with the max size and it looks like it might be a good prime suspect for our issues: it increases along with the used memory (makes sense) but it seems it never hits any threshold. Eventually all ODT can do seems to be garbage collection. 

![image](https://github.com/DependencyTrack/dependency-track/assets/129828933/200753da-3d8a-454e-8077-b4280d83015b)
![image](https://github.com/DependencyTrack/dependency-track/assets/129828933/ede590a4-92d0-41b1-94ba-d9e66e7ae2b7)
![image](https://github.com/DependencyTrack/dependency-track/assets/129828933/a4bad4ec-5fec-40a1-84f0-21db3727b443)
![image](https://github.com/DependencyTrack/dependency-track/assets/129828933/388503ec-2134-48d5-ad30-14b8ab3ded09)
![image](https://github.com/DependencyTrack/dependency-track/assets/129828933/53b28444-bbf0-4b68-8f19-2a54918c65d4)
![image](https://github.com/DependencyTrack/dependency-track/assets/129828933/ac8b2544-a725-4054-99a9-dd0acb9d74db)
![image](https://github.com/DependencyTrack/dependency-track/assets/129828933/c350ff7f-480d-4525-9ad7-6c5ce6bba9bb)
![image](https://github.com/DependencyTrack/dependency-track/assets/129828933/18a92340-a51b-483c-acac-4f1259e0be38)
![image](https://github.com/DependencyTrack/dependency-track/assets/129828933/8779c676-ddb0-4158-b050-c017b9c3aa89)

So our next move will be to try to configure the ALPINE_DATANUCLEUS_CACHE_LEVEL2_MAX_SIZE and see how it behaves, but will it be taken into account ? If not, do you have any thought regarding this issue ? As a reminder, we're sending cdxgen-generated SBOMs to ODT so that we can retrieve it afterwards with the vulnerabilities set. 

Thanks in advance for your insight !

**Reply by nscuro** [reply]
> We are limited by the kube cluster node's memory: it has a total of 32GB, but roughly 16GB are available (depending on the time of day due to other batches).
So basically when we increase the worker threads count and the memory, the server eventually crashes due to the lack of memory on the node.

What did you configure for `requests.memory` and `limits.memory`? If available memory changes during the day, and `requests.memory` is lower than `limits.memory`, then k8s might reduce the memory available to the DT pod, which can in turn lead to it getting OOMKilled, if it was consuming more than `requests.memory`. You can avoid this by setting `requests.memory` and `limits.memory` to an equal value. Here's a helpful article about his: https://xebia.com/blog/guide-kubernetes-jvm-integration/

> Increasing the memory to more than 16GB will be tricky due to our infra limitations. So we decreased it to 12GB and reduced the threads count accordingly.

Generally if you provide more memory to the JVM, it will use it. If there's no pressure to do garbage collection, it won't do it (per default, that can be tweaked via GC settings). So I would not recommend throwing more RAM at the problem - It will eventually fill up no matter how large it is.

Try aligning `requests.memory` with `limits.memory`. See if that helps. If not, try overriding [the default `MAXRamPercentage`](https://github.com/DependencyTrack/dependency-track/blob/e3217a373c110e817e4f4317af44c3c6840864ab/src/main/docker/Dockerfile#L18-L19) with a lower value:

```
extraEnv:
- name: EXTRA_JAVA_OPTIONS
  value: "-XX:MaxRAMPercentage=80.0"
```

The above will cause the JVM to allocate at most 80% of available RAM to its heap. That can help memory spaces outside of the JVM heap require more space.

You can also set this to an explicit value via `-Xmx`, e.g. `-Xmx10g`. Then, at most 10GB would be allocated to the JVM heap, of the 12GB you allocated to the pod.

> is the queue lost or do you maintain state in database ?

The event queue is entirely in-memory and thus will lose its state upon reboot or crash. It's one of the reasons we started working on Hyades (to which I linked in my first response).

> So our next move will be to try to configure the ALPINE_DATANUCLEUS_CACHE_LEVEL2_MAX_SIZE and see how it behaves, but will it be taken into account ?

All properties under `alpine.datanucleus` will be forwarded as-is to DataNucleus. It doesn't look like it supports a max cache size: https://www.datanucleus.org/products/accessplatform_6_0/jdo/persistence.html#cache_level2

**Reply by nekhtan** [reply]
You're amazingly fast to answer, that's awesome !!

> What did you configure for requests.memory and limits.memory?

On our last run, request and limit are the same, but they weren't previously.

> try overriding the default MAXRamPercentage with a lower value

We'll do that and keep you posted on how it behaves ;)

> The event queue is entirely in-memory and thus will lose its state upon reboot or crash. It's one of the reasons we started working on Hyades

That's what I thought, and yes we did look at Hyades (which looks great btw), but we think we're in too deep with v4 and business is expecting us to deliver soon... But rest assured that the plan is to migrate as soon as possible.

> It doesn't look like it supports a max cache size

That would be `datanucleus.cache.level2.maxSize` ([cache properties](https://www.datanucleus.org/products/accessplatform_6_0/jdo/persistence.html#pmf_props_dn_cache)), or did I miss something ? As I understand it, its default value is `-1` which means "no limit".

**Reply by nscuro** [reply]
> That would be datanucleus.cache.level2.maxSize ([cache properties](https://www.datanucleus.org/products/accessplatform_6_0/jdo/persistence.html#pmf_props_dn_cache)), or did I miss something ? 

Ha, you looked deeper into the docs than me. TIL that property exists! :D

If it turns out to work for you, perhaps we should add this to the docs as another knob to tweak when memory usage gets too high...

**Reply by nscuro** [reply]
If you're feeling adventurous:

In Hyades we have switched to the G1 garbage collector and enabled String deduplication. We found that G1 per default does a better job of keeping the heap manageable, while also having shorter pause times:

https://github.com/DependencyTrack/hyades-apiserver/blob/40dfe3b1da0aadc52aba508c4effc057171a618d/src/main/docker/Dockerfile#L35C19-L35C84

Saying this because I just now saw the average 8s GC pause times you're experiencing under load. That's **a lot**.

**Reply by nekhtan** [reply]
Hello @nscuro 

I didn't try yet the G1GC but it's definitely on my todo list.

Since last time I tried to configure the datanucleus' `maxSize`. It works well on startup but not so much under high pressure... There's a TODO in the [putAll](https://github.com/datanucleus/datanucleus-core/blob/master/src/main/java/org/datanucleus/cache/AbstractReferencedLevel2Cache.java#L295) method so depending on how it's being used it might have no effect to configure it. I'm not saying it's the case here, but right now I can't explain why setting a limit to 2M is ignored while setting a limit to 2k is taken into account.

I also tried to completely disable the cache (`alpine.datanucleus.cache.level2.type=none`) it looked ok overall (even though it was queuing more) but it was on the end of a run, so I might try that again later on a full run. That being said I'd really prefer not disable it entirely.

That's why I eventually tried with the weak references cache (`alpine.datanucleus.cache.level2.type=weak`) and it looks promising, even though you're [advising against changing it](https://docs.dependencytrack.org/getting-started/configuration/):

![image](https://github.com/DependencyTrack/dependency-track/assets/129828933/cf6692d3-aee5-4f5a-a668-4376ffb06e69)

The database operations / queries have slightly increased compared to a run with the soft cache while the cache is MUCH smaller since its values can be GCed. So it makes me wonder, are the cached entries really that reusable or do you expect it to be just a small subset of data that are actually queried more often than the vast majority?

**Reply by nscuro** [reply]
> are the cached entries really that reusable or do you expect it to be just a small subset of data that are actually queried more often than the vast majority?

That pretty much hits the nail on its head. That is exactly what we have observed. The problem is that L2 caching is enabled globally per default, so *everything* ends up being put into the cache as soon as it's "touched" by the code.

But in order to get a cache hit, objects need to be queried by their ID (i.e. the `ID` database column). But DT mostly uses UUIDs (which is separate from the primary key) in its API. Generally *most* queries are not being performed based on ID, but some other fields instead. So the L2 cache ends up being useless for *a lot* of objects.

Usually you'd enable L2 caching for the areas where you *know* you need it, making it easier to control cache sizes. We did end up in the reverse situation, which makes it a bit harder to resolve.

I have been working on this for quite a while now, in the sense that for all new features, and code areas I am refactoring, I intentionally *disable* L2 caching. So one step after another, I'm positive we can drastically decrease the negative impact it has.

Sometime ago I did experiment with changing the L2 cache type to `weak`, but as you already figured out, it might be GCing entries *too* soon, to a point where you don't get much benefit out of the caching to begin with.

In our experience, G1 is a bit more aggressive with cleaning `soft` references, whereas the default ParallelGC tends to keep them around for longer.
